worker/controlsocket/worker.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+//go:build linux
+
+package controlsocket
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/juju/sockets"
+)
+
+// logger is here to stop the desire of creating a package level logger.
+// Don't do this, instead use the one passed as manifold config.
+type logger interface{}
+
+var _ logger = struct{}{}
+
+// Logger represents the methods used by the worker to log information.
+type Logger interface {
+	Errorf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+	Debugf(string, ...interface{})
+	Tracef(string, ...interface{})
+}
+
+// Config represents configuration for the controlsocket worker.
+type Config struct {
+	State      State
+	Logger     Logger
+	SocketName string
+}
+
+// Validate returns an error if config cannot drive the Worker.
+func (config Config) Validate() error {
+	if config.State == nil {
+		return errors.NotValidf("nil state")
+	}
+	if config.Logger == nil {
+		return errors.NotValidf("nil logger")
+	}
+	if config.SocketName == "nil" {
+		return errors.NotValidf("empty socketName")
+	}
+	return nil
+}
+
+// Worker is a controlsocket worker.
+type Worker struct {
+	Config
+	tomb     tomb.Tomb
+	listener net.Listener
+}
+
+// NewWorker returns a controlsocket worker with the given config.
+func NewWorker(config Config) (worker.Worker, error) {
```

Comment:
Worker needs tests.

Reply:
what kind of tests? All it does is put up an HTTP server, and I've already tested the handlers to death.

Reply:
Doesn't need to test what it is doing, just that can startup and shutdown cleanly.